keps/sig-network/20200617-NodeIPAM-controller-support-variable-sized-CIDRs/README.md

```diff
@@ -0,0 +1,645 @@
+motivation section can optionally provide links to [experience reports][] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+Some user want to change `--node-cidr-mask-size` after create kubernetes cluster by some reason, 
+but current kubernetes doesn't allow `--node-cidr-mask-size` to change, because NodeIPAM controller
+doesn't support variable sized CIDRs.
+
+
+
+### Goals
+
+<!--
+List the specific goals of the KEP.  What is it trying to achieve?  How will we
+know that this has succeeded?
+-->
+
+Make NodeIPAM controller support variable sized CIDRs.
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP?  Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation.  The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+Use interval tree instead of bitmap to allocate podCIDR.
+
+### User Stories (optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system.  The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+For example, we have a kubernetes cluster, the `clusterCIDR` is 192.168.0.0/16 and `--node-cidr-mask-size` is 24
+1. add two nodes to cluster, podCIDR will be `192.168.0.0/24` and `192.168.1.0/24`;
+2. change `--node-cidr-mask-size` to 23 and delete node `192.168.1.0/24`;
+3. add a node pod CIDR will be `192.168.0.0/23`, `192.168.0.0/24` and `192.168.0.0/23` overlap.
+
+#### Story 2
+
+### Notes/Constraints/Caveats (optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above.
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal and how do we mitigate.  Think broadly.
+For example, consider both security and how this will impact the larger
+kubernetes ecosystem.
+
+How will security be reviewed and by whom?
+
+How will UX be reviewed and by whom?
+
+Consider including folks that also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable.  This may include API specs (though not always
+required) or even code snippets.  If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+We need use a new data structure instead of bitmap to allocate podCIDR, it should support variable sized CIDRs, 
```

Comment:
What I was really looking for is a plan for how you can consume and existing bitmaps and convert to the new struct.  But in digging, it seems we don't actually STORE the allocated CIDRs anywhere, so maybe you're lucky and this is easier than I thought?

Reply:
yeah, the bitmap is not persisted as in the allocators, the data is persisted in the Node object
https://github.com/kubernetes/kubernetes/blob/323f34858de18b862d43c40b2cced65ad8e24052/pkg/controller/nodeipam/ipam/range_allocator.go#L226-L238

Reply:
:thinking: , how will work 2 controllers with different cidr_set implementations at the same time?
at least the old one will not be able to do resizing o cirds

Reply:
Yeah, we'll need a very well thought out deployment plan.  Specifically, covering cases like:

* Have 3-controller-manager instances
* Upgrade 1 of them to this new model
* Allocate node with the new CID size
* Leader election changes which controller-manager is in charge

During that period, some masters will still be using bitmaps.  That is going to end badly. 

We'd need at least a multi-stage rollout, where we convert controllers to use intervals but not change the config.  Once that is rolled out you can change the config.

But you also need to handle rollbacks, so you can't go back into a corrupt state.

Finally, is it really what people want to have a fixed CIDR size?  Maybe we should be shooting for something more policy oriented?